src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/DateTimeEqualityRewriter.cs

```diff
@@ -0,0 +1,93 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Expressions
+{
+    /// <summary>
+    /// Rewrites (And (FieldGreaterThanOrEqual DateTimeStart x) (FieldLessThanOrEqual DateTimeEnd y)) to
+    /// (And (FieldGreaterThanOrEqual DateTimeStart x) (FieldLessThanOrEqual DateTimeEnd y) (FieldLessThanOrEqual DateTimeStart y)).
+    /// It looks specifically for this pattern because it is what we emit in the core layer for an equals search over a date parameter.
+    /// This rewriting constrains the range scan over the index (DateTimeStart, DateTimeEnd).
+    /// </summary>
+    internal class DateTimeEqualityRewriter : ExpressionRewriterWithInitialContext<object>
+    {
+        internal static readonly DateTimeEqualityRewriter Instance = new();
+
+        public override Expression VisitSearchParameter(SearchParameterExpression expression, object context)
+        {
+            if (expression.Parameter.Type == SearchParamType.Date ||
+                (expression.Parameter.Type == SearchParamType.Composite && expression.Parameter.ResolvedComponents.Any(c => c.Type == SearchParamType.Date)))
+            {
+                return base.VisitSearchParameter(expression, context);
+            }
+
+            return expression;
+        }
+
+        public override Expression VisitMultiary(MultiaryExpression expression, object context)
+        {
+            expression = (MultiaryExpression)base.VisitMultiary(expression, context);
+            if (expression.MultiaryOperation != MultiaryOperator.And)
+            {
+                return expression;
+            }
+
+            List<Expression> newExpressions = null;
+            int i = 0;
+            for (; i < expression.Expressions.Count - 1; i++)
+            {
+                switch (MatchPattern(expression.Expressions[i], expression.Expressions[i + 1]))
+                {
+                    case ({ } low, { } high):
```

Comment:
>{ } low, { } high [](start = 26, length = 17)

I never saw this syntax before. What these curly braces do?

Reply:
It matches non-null values. Alternative would be `(BinaryExpression low, BinaryExpression high)`. But `(var low, var high)` matches null values which is pretty surprising design decision IMO.